library/eddsa.c

```diff
@@ -0,0 +1,446 @@
+        unsigned char sha_buf[64], tmp_buf[32];
+        size_t olen = 0;
+
+        /* r computation */
+        mbedtls_sha512_init( &sha_ctx );
+        mbedtls_sha512_starts( &sha_ctx , 0 );
+
+        if( eddsa_id == MBEDTLS_EDDSA_CTX )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 0, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+        else if( eddsa_id == MBEDTLS_EDDSA_PREHASH )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 1, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+
+        /* Update SHA with prefix */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary_le( &prefix, tmp_buf, sizeof(tmp_buf) ) );
+
+        mbedtls_sha512_update( &sha_ctx, tmp_buf, sizeof( tmp_buf ) );
+
+        /* In EDDSA_PREHASH, buf should contain the SHA512 hash. It contains the whole message otherwise */
+        mbedtls_sha512_update( &sha_ctx, buf, blen );
+
+        mbedtls_sha512_finish( &sha_ctx, sha_buf );
+        mbedtls_sha512_free( &sha_ctx );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary_le( &rq, sha_buf, sizeof( sha_buf ) ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &rq, &rq, &grp->N ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_mul( grp, &R, &rq, &grp->G, f_rng, p_rng ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_copy( r, &R.Y ) );
+        if( mbedtls_mpi_get_bit( &R.X, 0 ) )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_mpi_set_bit( r, 255, 1 ) );
+        }
+
+        /* s computation */
+        mbedtls_sha512_init( &sha_ctx );
+        mbedtls_sha512_starts( &sha_ctx , 0 );
+
+        if( eddsa_id == MBEDTLS_EDDSA_CTX )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 0, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+        else if( eddsa_id == MBEDTLS_EDDSA_PREHASH )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 1, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_point_write_binary( grp, &R, MBEDTLS_ECP_PF_COMPRESSED, &olen, tmp_buf, sizeof( tmp_buf ) ) );
+        mbedtls_sha512_update( &sha_ctx, tmp_buf, sizeof( tmp_buf ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_point_write_binary( grp, &Q, MBEDTLS_ECP_PF_COMPRESSED, &olen, tmp_buf, sizeof( tmp_buf ) ) );
+        mbedtls_sha512_update( &sha_ctx, tmp_buf, sizeof( tmp_buf ) );
+
+        /* In EDDSA_PREHASH, buf should contain the SHA512 hash. It contains the whole message otherwise */
+        mbedtls_sha512_update( &sha_ctx, buf, blen );
+
+        mbedtls_sha512_finish( &sha_ctx, sha_buf );
+        mbedtls_sha512_free( &sha_ctx );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary_le( &h, sha_buf, sizeof( sha_buf ) ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &h, &h, &grp->N ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( &h, &h, &q ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_add_mpi( s, &h, &rq ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( s, s, &grp->N ) );
+    }
+#endif
+
+cleanup:
```

Comment:
`tmp_buf` contains the ephemeral key at some point. You should use `mbedtls_platform_zeroize()` to make sure it is securily erased.